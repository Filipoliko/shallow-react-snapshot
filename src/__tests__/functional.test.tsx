import React from 'react';
import { createPortal } from 'react-dom';
import { shallow } from '../index';
import { render } from '@testing-library/react';

function MyComponent({ children, ...props }: { children?: React.ReactNode, [key: string]: any }) {
    return <div id="MyComponent" {...props}>{children || null}</div>;
}

function MyComponentWithChildrenOnSecondPosition({ children }: { children: React.ReactNode }) {
    return (
      <div id="MyComponentWithChildrenOnSecondPosition">
        <div>First</div>
        {children || null}
      </div>
    );
}
  
const MyMemoizedComponent = React.memo(MyComponent);
const MyForwardedComponent = React.forwardRef(function MyForwardedComponent({ children }: { children: React.ReactNode }, ref: React.Ref<HTMLDivElement>) {
    return <div id="MyForwardedComponent" ref={ref}>{children}</div>;
});

const testCases = [
  {
    description: 'direct wrapper around react component with no child element',
    App: function App() {
      return <MyComponent />;
    },
  },
  {
    description: 'direct wrapper around react component with a single child element',
    App: function App() {
      return (
        <MyComponent>
          <h1>Hello World One</h1>
        </MyComponent>
      );
    },
  },
  {
    description: 'direct wrapper around react component with a single child element and children on second position',
    App: function App() {
      return (
        <MyComponentWithChildrenOnSecondPosition>
          <h1>Hello World One</h1>
        </MyComponentWithChildrenOnSecondPosition>
      );
    },
  },
  {
    description: 'direct wrapper around react component with a single child element and props',
    App: function App() {
      return (
        <MyComponent something={'beautiful'}>
          <h1>Hello World One</h1>
        </MyComponent>
      );
    },
  },
  {
    description: 'direct wrapper around react component with multiple deep child elements',
    App: function App() {
      return (
        <MyComponent>
          <MyComponent>
            <h1>Hello World One</h1>
          </MyComponent>
        </MyComponent>
      );
    },
  },
  {
    description: 'wrapper around react component with multiple children',
    App: function App() {
      return (
        <div>
          <MyComponent>
            <h2>Hello World Two</h2>
            Some random text
          </MyComponent>
        </div>
      );
    },
  },
  {
    description: 'wrapper around react component with single text child',
    App: function App() {
      return (
        <div>
          <MyComponent>
            Some random text
          </MyComponent>
        </div>
      );
    },
  },
  {
    description: 'wrapper around react component with single number child',
    App: function App() {
      return (
        <div>
          <MyComponent>
            {1234}
          </MyComponent>
        </div>
      );
    },
  },
  {
    description: 'fragment wrapper around react component with single text child',
    App: function App() {
      return (
        <>
          <MyComponent>
            <>Some random text</>
          </MyComponent>
        </>
      );
    },
  },
  {
    description: 'direct memoized wrapper around react component with a single child element',
    App: function App() {
      return (
        <MyMemoizedComponent>
          <h1>Hello World One</h1>
        </MyMemoizedComponent>
      );
    },
  },
  {
    description: 'direct forwarded wrapper around react component with a single child element',
    App: function App() {
      return (
        <MyForwardedComponent>
          <h1>Hello World One</h1>
        </MyForwardedComponent>
      );
    },
  },
  {
    description: 'direct profiler wrapper around react component with a single child element',
    App: function App() {
      return (
        <React.Profiler id="profiler" onRender={() => {}}>
          <MyComponent>
            <h1>Hello World One</h1>
          </MyComponent>
        </React.Profiler>
      );
    },
  },
  {
    description: 'direct StrictMode wrapper around react component with a single child element',
    App: function App() {
      return (
        <React.StrictMode>
          <MyComponent>
            <h1>Hello World One</h1>
          </MyComponent>
        </React.StrictMode>
      );
    },
  },
// @TODO: Suspense is rendering a bit differently for React 16
//   {
//     description: 'direct suspense wrapper around react component with a single child element',
//     App: function App() {
//       return (
//         <React.Suspense fallback={<div>Loading...</div>}>
//           <MyComponent>
//             <h1>Hello World One</h1>
//           </MyComponent>
//         </React.Suspense>
//       );
//     },
//   },
  {
    description: 'react component with portal',
    App: function App() {
      return (
        <MyComponent>
          {createPortal(<h1>Hello World One</h1>, document.body)}
        </MyComponent>
      );
    },
  },
];

describe('Functional component render', () => {
    test.each(testCases)('$description', ({ App }) => {
        const { container } = render(<App />);

        const result = shallow(container.firstChild as HTMLElement, { whitelist: [App] });

        expect(result).toMatchSnapshot();
    });
});
