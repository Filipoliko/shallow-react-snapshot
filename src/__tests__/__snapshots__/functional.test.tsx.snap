// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Functional component render direct StrictMode wrapper around react component with a single child element 1`] = `
<StrictMode>
  <MyComponent>
    <h1>
      Hello World One
    </h1>
  </MyComponent>
</StrictMode>
`;

exports[`Functional component render direct forwarded wrapper around react component with a single child element 1`] = `
<ForwardRef(MyForwardedComponent)>
  <h1>
    Hello World One
  </h1>
</ForwardRef(MyForwardedComponent)>
`;

exports[`Functional component render direct memoized wrapper around react component with a single child element 1`] = `
<MyComponent>
  <h1>
    Hello World One
  </h1>
</MyComponent>
`;

exports[`Functional component render direct profiler wrapper around react component with a single child element 1`] = `
<Profiler
  id="profiler"
  onRender={[Function]}
>
  <MyComponent>
    <h1>
      Hello World One
    </h1>
  </MyComponent>
</Profiler>
`;

exports[`Functional component render direct wrapper around react component with a single child element 1`] = `
<MyComponent>
  <h1>
    Hello World One
  </h1>
</MyComponent>
`;

exports[`Functional component render direct wrapper around react component with a single child element and children on second position 1`] = `
<MyComponentWithChildrenOnSecondPosition>
  <h1>
    Hello World One
  </h1>
</MyComponentWithChildrenOnSecondPosition>
`;

exports[`Functional component render direct wrapper around react component with a single child element and props 1`] = `
<MyComponent
  something="beautiful"
>
  <h1>
    Hello World One
  </h1>
</MyComponent>
`;

exports[`Functional component render direct wrapper around react component with multiple deep child elements 1`] = `
<MyComponent>
  <MyComponent>
    <h1>
      Hello World One
    </h1>
  </MyComponent>
</MyComponent>
`;

exports[`Functional component render direct wrapper around react component with no child element 1`] = `<MyComponent />`;

exports[`Functional component render fragment wrapper around react component with single text child 1`] = `
<MyComponent>
  <Fragment>
    Some random text
  </Fragment>
</MyComponent>
`;

exports[`Functional component render react component with portal 1`] = `
<MyComponent>
  <h1>
    Hello World One
  </h1>
</MyComponent>
`;

exports[`Functional component render wrapper around react component with multiple children 1`] = `
<div>
  <MyComponent>
    <h2>
      Hello World Two
    </h2>
    Some random text
  </MyComponent>
</div>
`;

exports[`Functional component render wrapper around react component with single number child 1`] = `
<div>
  <MyComponent>
    1234
  </MyComponent>
</div>
`;

exports[`Functional component render wrapper around react component with single text child 1`] = `
<div>
  <MyComponent>
    Some random text
  </MyComponent>
</div>
`;
